src/Juvix/Backends/Michelson/Compilation/Term.hs

```diff
@@ -7,167 +7,276 @@ import Juvix.Backends.Michelson.Compilation.Prim
-    J.App func arg →
-      stackCheck addsOne $ do
-        func ← termToInstr func paramTy -- :: Lam a b
-        arg ← termToInstr arg paramTy -- :: a
-        modify @"stack"
-          ( \(_ : (_, (M.Type (M.TLambda _ retTy) _)) : xs) →
-              (FuncResultE, retTy) : xs
-          )
-        pure
-          ( M.SeqEx
-              [ func, -- Evaluate the function.
-                arg, -- Evaluate the argument.
-                M.PrimEx (M.EXEC "") -- Execute the function.
-              ]
-          )
+    J.App _ _ → do
+      -- TODO: figure out how to do stack check here
+      let (lam, args) = argsFromApps ann
+      case lam of
+        (J.LamM captures arguments body, _usage, lamTy) → do
+          insts ← evaluateAndPushArgs arguments lamTy args paramTy
+          recurseApplication (captures, arguments, body) lamTy args insts paramTy
+        ann@(J.Prim prim, _, primTy) → do
+          -- Treat the primitive as a function with n arguments, the body will eventually be inlined (or packed in a lambda).
+          let arguments = replicate (arity prim) "_"
+          insts ← evaluateAndPushArgs arguments primTy args paramTy
+          recurseApplication ([], arguments {- the arguments to the primitive -}, ann {- will be inlined -}) primTy args insts paramTy
+        t → do
+          failWith ("Applications applied to non lambda term: " <> show t)
+
+-- captures, arguments, body are for the function
+-- lamTy is the type of the function
+-- args are the arguments to which it is being applied
+recurseApplication ∷
+  ∀ m a.
+  ( HasState "stack" VStack.T m,
+    HasThrow "compilationError" CompilationError m,
+    HasWriter "compilationLog" [CompilationLog] m
+  ) ⇒
+  ([Symbol], [Symbol], Term) →
+  J.Type PrimTy PrimVal →
+  [a] →
+  [Op] →
+  M.Type →
+  m (Either LamPartial M.ExpandedOp)
+recurseApplication (captures, lamArguments, body) lamTy args insts paramTy = do
+  let argsL = length args
+      lamArgsL = length lamArguments
+  case compare argsL lamArgsL of
+    EQ → do
+      -- Exactly applied case, we have the correct number of arguments.
+      -- need to assert that we correctly drop them later, this is complicated!
+      -- need to pass things to be dropped along with the partially applied function datatype
+      eitherFuncOp ← termToInstr body paramTy
+      case eitherFuncOp of
+        Right f → do
+          -- Fully applied case without a lambda: evaluate all the arguments, inline the function, drop the args.
+          pure $
+            Right
+              ( M.SeqEx
+                  ( insts
+                      <> [ f,
+                           M.PrimEx
+                             (M.DIP (replicate argsL (M.PrimEx M.DROP)))
+                         ]
+                  )
+              )
+        Left (LamPartial ops captures remArgs body ty) → do
+          -- Fully applied case with a returned lambda: return the instructions followed by the lambda, defer evaluation of the body.
```

Comment:
Here is where we might need to worry about deleting the arguments, since we evaluate them (`insts`) but don't necessarily drop them anywhere - I wonder if we need to track this on `LamPartial` (how many arguments to drop after inlining the body)